orttraining/orttraining/training_ops/cuda/optimizer/adam/adam.cc

```diff
@@ -0,0 +1,148 @@
+  size_t num_of_gradients = gradients->Size();
+  size_t num_of_momentums_1 = momentums_1->Size();
+  size_t num_of_momentums_2 = momentums_2->Size();
+
+  // Check the number of weights, gradients, momentums matchs.
+  ORT_ENFORCE(num_of_weights > 0, "invalid count of tensors in Seq<Tensor>.");
+  ORT_ENFORCE(num_of_weights == num_of_gradients, "number of weights and gradients mismatch.");
+  ORT_ENFORCE(num_of_gradients == num_of_momentums_1, "number of gradients and momentums_1 mismatch.");
+  ORT_ENFORCE(num_of_momentums_1 == num_of_momentums_2, "number of momentums_1 and momentums_2 mismatch.");
+
+  std::vector<int> tensor_sizes(num_of_weights);
+  std::vector<std::vector<void*>> grouped_tensor_pointers(num_of_weights);
+
+  for (size_t i = 0; i < num_of_weights; ++i) {
+    const Tensor& weight_tensor = weights->Get(i);
+    const Tensor& gradient_tensor = gradients->Get(i);
+    const Tensor& momentum_1_tensor = momentums_1->Get(i);
+    const Tensor& momentum_2_tensor = momentums_2->Get(i);
+
+    // Check the weight/gradient/momentums at the same index should have same shape.
+    ORT_ENFORCE(weight_tensor.Shape() == gradient_tensor.Shape(), "shape of weight and gradient mismatch.");
+    ORT_ENFORCE(gradient_tensor.Shape() == momentum_1_tensor.Shape(), "shape of gradient and momentum_1 mismatch.");
+    ORT_ENFORCE(momentum_1_tensor.Shape() == momentum_2_tensor.Shape(), "shape of momentum_1 and momentum_2 mismatch.");
+
+    // Currently we only support float data types.
+    ORT_ENFORCE(weight_tensor.IsDataType<float>() &&
+                    gradient_tensor.IsDataType<float>() &&
+                    momentum_1_tensor.IsDataType<float>() &&
+                    momentum_2_tensor.IsDataType<float>(),
+                "Only float data type support for Adam cuda kernel.");
+
+    tensor_sizes[i] = static_cast<int>(weight_tensor.Shape().Size());
+
+    grouped_tensor_pointers[i] = {
+        const_cast<float*>(weight_tensor.Data<float>()),
+        const_cast<float*>(gradient_tensor.Data<float>()),
+        const_cast<float*>(momentum_1_tensor.Data<float>()),
+        const_cast<float*>(momentum_2_tensor.Data<float>())};
+  }
+
+  Tensor* updated_flag = ctx->Output(0, step->Shape());
+  TensorSeq* updated_weights = ctx->Output<TensorSeq>(1);
+  TensorSeq* updated_momentums_1 = ctx->Output<TensorSeq>(2);
+  TensorSeq* updated_momentums_2 = ctx->Output<TensorSeq>(3);
+
+  typedef typename ToCudaType<float>::MappedType CudaT_FLOAT;
+  typedef AdamMTAFunctor<CudaT_FLOAT, CudaT_FLOAT, CudaT_FLOAT> TFunctor;
+  TFunctor functor;
+
+  const float* lr_ptr = learning_rate->template Data<float>();
+  const int64_t* step_ptr = step->template Data<int64_t>();
+  ORT_ENFORCE(lr_ptr && step_ptr);
+
+  int64_t* updated_flag_ptr = updated_flag->template MutableData<int64_t>();
+  *updated_flag_ptr = 1;
+
+  launch_multi_tensor_functor<MTA_ADAM_GROUP_SIZE, TFunctor>(
+      Stream(), 2048 * 32, tensor_sizes, grouped_tensor_pointers, functor,
```

Comment:
How do we come up with `2048*32` as the `chunk_size`? Shall we put it in a `contsexpr` if possible?

Reply:
I am following the one LambOptimizer and FusedAdam used, yeah let me make it a const.